File: frontend/src/api/wizard.ts
/**
 * 向导模块 API
 * 支持创作向导的 SSE 流式调用
 * 支持新的地图-部分-章节层级结构
 */
import { getAccessToken } from './request'

// ============ UUID 生成工具 ============

export function generateUUID(): string {
  return crypto.randomUUID ? crypto.randomUUID().slice(0, 8) :
    Math.random().toString(36).slice(2, 10)
}

// ============ 新的数据结构 ============

export interface SceneNode {
  id?: string
  name: string
  description?: string
}

export interface ChapterOutlineItem {
  id?: string
  chapter: number
  title: string
  summary: string
  scenes?: SceneNode[]
}

export interface PartNode {
  id?: string
  name: string
  summary?: string
  chapters: ChapterOutlineItem[]
  character_ids: string[]
}

export interface MapNode {
  id?: string
  name: string
  description?: string
  parts: PartNode[]
}

export interface CharacterOutlineItem {
  id?: string
  name: string
  role_type: string
  gender?: string
  age?: string
  occupation?: string
  personality_traits?: string
  appearance?: string
  background?: string
  appearances?: string[]
  origin_map_id?: string  // 初次登场的地图ID
  is_new?: boolean
}

export interface NoteItem {
  id?: string
  note_type: 'foreshadowing' | 'inspiration' | 'note'
  title: string
  content?: string
  related_chapter_ids?: string[]
  target_type?: 'map' | 'part' | 'chapter'  // 关联目标类型
  target_id?: string  // 关联目标ID
  status?: 'active' | 'resolved' | 'abandoned'
}

// ============ 新的请求/响应类型 ============

export interface WizardMapsRequest {
  title: string
  genre?: string
  description: string
  reference_ids?: number[]
  revision_request?: string
  current_maps?: MapNode[]
}

export interface WizardPartsRequest {
  title: string
  genre?: string
  description: string
  map_id: string
  map_name: string
  revision_request?: string
  current_parts?: PartNode[]
}

export interface WizardCharactersForPartRequest {
  title: string
  genre?: string
  description: string
  parts: PartNode[]
  existing_characters?: CharacterOutlineItem[]
}

export interface WizardCreateV2Request {
  title: string
  genre?: string
  description?: string
  maps: MapNode[]
  characters: CharacterOutlineItem[]
  notes?: NoteItem[]
  reference_ids?: number[]
}

export interface WizardCreateResponse {
  project_id: number
  message: string
}

// ============ 旧接口类型（保留兼容） ============

export interface WizardGenerateRequest {
  title: string
  genre?: string
  description: string
  target_word_count: number
  chapter_count: number
  reference_ids?: number[]
  revision_request?: string
  current_outline?: ChapterOutlineItem[]
  current_characters?: CharacterOutlineItem[]
}

export interface WizardOutlineRequest {
  title: string
  genre?: string
  description: string
  target_word_count: number
  chapter_count: number
  reference_ids?: number[]
  revision_request?: string
  current_outline?: ChapterOutlineItem[]
}

export interface WizardCharactersRequest {
  title: string
  genre?: string
  description: string
  outline: ChapterOutlineItem[]
}

export interface WizardCreateRequest {
  title: string
  genre?: string
  description?: string
  target_word_count: number
  outline: ChapterOutlineItem[]
  characters: CharacterOutlineItem[]
  reference_ids?: number[]
  outline_text?: string
}

// ============ SSE 事件类型 ============

export interface WizardGenerateEvent {
  type: 'progress' | 'outline' | 'characters' | 'maps' | 'parts' | 'done' | 'error'
  message?: string
  data?: any
}

// ============ SSE 流式调用工具函数 ============

function createSSEStream(
  url: string,
  data: any,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  const controller = new AbortController()

  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  const token = getAccessToken()
  if (token) {
    headers['Authorization'] = `Bearer ${token}`
  }

  fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(data),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        const err = await response.json().catch(() => ({ detail: '请求失败' }))
        onError(err.detail || '请求失败')
        return
      }

      const reader = response.body?.getReader()
      if (!reader) {
        onError('无法读取响应流')
        return
      }

      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const payload = JSON.parse(line.slice(6)) as WizardGenerateEvent
              if (payload.type === 'error') {
                onError(payload.message || '生成失败')
                return
              }
              onEvent(payload)
            } catch {
              // 忽略解析错误
            }
          }
        }
      }
    })
    .catch((err) => {
      if (err.name !== 'AbortError') {
        onError(err.message || '网络请求失败')
      }
    })

  return controller
}

// ============ 新的 API 函数 ============

/**
 * 步骤2：生成地图大纲（SSE 流式）
 */
export function streamWizardMaps(
  data: WizardMapsRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate-maps', data, onEvent, onError)
}

/**
 * 步骤3：为地图生成部分（SSE 流式）
 */
export function streamWizardParts(
  data: WizardPartsRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate-parts', data, onEvent, onError)
}

/**
 * 步骤4：为部分生成角色（SSE 流式）
 */
export function streamWizardCharactersForPart(
  data: WizardCharactersForPartRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate-characters-for-part', data, onEvent, onError)
}

/**
 * 创建向导项目（新版本）
 */
export async function createWizardProjectV2(data: WizardCreateV2Request): Promise<WizardCreateResponse> {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  const token = getAccessToken()
  if (token) {
    headers['Authorization'] = `Bearer ${token}`
  }

  const resp = await fetch('/api/v1/wizard/create-v2', {
    method: 'POST',
    headers,
    body: JSON.stringify(data),
  })
  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: '创建失败' }))
    throw new Error(err.detail || '创建失败')
  }
  return resp.json()
}

// ============ 旧 API 函数（保留兼容） ============

export function streamWizardGenerate(
  data: WizardGenerateRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate', data, onEvent, onError)
}

export function streamWizardOutline(
  data: WizardOutlineRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate-outline', data, onEvent, onError)
}

export function streamWizardCharacters(
  data: WizardCharactersRequest,
  onEvent: (event: WizardGenerateEvent) => void,
  onError: (error: string) => void,
): AbortController {
  return createSSEStream('/api/v1/wizard/generate-characters', data, onEvent, onError)
}

export async function createWizardProject(data: WizardCreateRequest): Promise<WizardCreateResponse> {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  const token = getAccessToken()
  if (token) {
    headers['Authorization'] = `Bearer ${token}`
  }

  const resp = await fetch('/api/v1/wizard/create', {
    method: 'POST',
    headers,
    body: JSON.stringify(data),
  })
  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: '创建失败' }))
    throw new Error(err.detail || '创建失败')
  }
  return resp.json()
}

File: backend/app/schemas/wizard.py
"""
向导模块 Pydantic 模式
用于创作向导的请求和响应定义
支持新的地图->部分->章节层级结构
"""
from __future__ import annotations
from typing import List, Optional, Union
import uuid

from pydantic import BaseModel, ConfigDict, field_validator


def generate_uuid() -> str:
    """生成 UUID"""
    return str(uuid.uuid4())[:8]


def coerce_id(v: Union[str, int, None]) -> Optional[str]:
    """将 ID 转换为字符串（兼容前端发送整数的情况）"""
    if v is None:
        return None
    return str(v)


# ============ 新的数据结构 ============

class SceneNode(BaseModel):
    """场景节点"""
    id: Optional[str] = None
    name: str
    description: Optional[str] = None

    _validate_id = field_validator('id', mode='before')(coerce_id)


class ChapterOutlineItem(BaseModel):
    """章节大纲项"""
    id: Optional[str] = None
    chapter: int
    title: str
    summary: str
    scenes: Optional[List[SceneNode]] = None

    _validate_id = field_validator('id', mode='before')(coerce_id)


class PartNode(BaseModel):
    """部分节点"""
    id: Optional[str] = None
    name: str  # 第一部分、第二部分
    summary: Optional[str] = None
    chapters: List[ChapterOutlineItem] = []
    character_ids: List[str] = []  # 出场角色ID

    _validate_id = field_validator('id', mode='before')(coerce_id)


class MapNode(BaseModel):
    """地图节点"""
    id: Optional[str] = None
    name: str  # 青云宗、北境荒漠
    description: Optional[str] = None
    parts: List[PartNode] = []

    _validate_id = field_validator('id', mode='before')(coerce_id)


class CharacterOutlineItem(BaseModel):
    """角色大纲项"""
    id: Optional[str] = None
    name: str
    role_type: str = "supporting"  # protagonist/antagonist/supporting/minor
    gender: Optional[str] = None
    age: Optional[str] = None
    occupation: Optional[str] = None
    personality_traits: Optional[str] = None
    appearance: Optional[str] = None
    background: Optional[str] = None
    appearances: List[str] = []  # 出场章节ID列表
    origin_map_id: Optional[str] = None  # 初次登场的地图ID

    _validate_id = field_validator('id', 'origin_map_id', mode='before')(coerce_id)


class NoteItem(BaseModel):
    """笔记项"""
    id: Optional[str] = None
    note_type: str = "note"  # foreshadowing/inspiration/note
    title: str
    content: Optional[str] = None
    related_chapter_ids: List[str] = []
    target_type: Optional[str] = None  # map/part/chapter 关联目标类型
    target_id: Optional[str] = None  # 关联目标ID
    status: str = "active"  # active/resolved/abandoned

    _validate_id = field_validator('id', 'target_id', mode='before')(coerce_id)


# ============ 新的向导请求/响应 ============

class WizardIdeaRequest(BaseModel):
    """步骤1：创作思路请求（简化版）"""
    title: str
    genre: Optional[str] = None
    description: str
    reference_ids: Optional[List[int]] = None


class WizardMapsRequest(BaseModel):
    """步骤2：生成地图请求"""
    title: str
    genre: Optional[str] = None
    description: str
    reference_ids: Optional[List[int]] = None
    # 修改意见
    revision_request: Optional[str] = None
    current_maps: Optional[List[MapNode]] = None


class WizardPartsRequest(BaseModel):
    """步骤3：生成部分请求"""
    title: str
    genre: Optional[str] = None
    description: str
    map_id: str  # 选中的地图ID
    map_name: str  # 地图名称
    # 修改意见
    revision_request: Optional[str] = None
    current_parts: Optional[List[PartNode]] = None


class WizardCharactersForPartRequest(BaseModel):
    """步骤4：为部分生成角色请求"""
    title: str
    genre: Optional[str] = None
    description: str
    parts: List[PartNode]  # 已确认的部分
    existing_characters: Optional[List[CharacterOutlineItem]] = None  # 已有角色库


class WizardCreateV2Request(BaseModel):
    """向导创建项目请求（新版本）"""
    title: str
    genre: Optional[str] = None
    description: Optional[str] = None
    maps: List[MapNode]  # 地图结构
    characters: List[CharacterOutlineItem]  # 角色库
    notes: Optional[List[NoteItem]] = None  # 笔记（可选）
    reference_ids: Optional[List[int]] = None


# ============ 保留旧接口兼容 ============

class WizardGenerateRequest(BaseModel):
    """向导生成请求：生成大纲和角色（保留兼容性）"""
    title: str
    genre: Optional[str] = None
    description: str
    target_word_count: int = 100000
    chapter_count: int = 10
    reference_ids: Optional[List[int]] = None
    revision_request: Optional[str] = None
    current_outline: Optional[List[ChapterOutlineItem]] = None
    current_characters: Optional[List[CharacterOutlineItem]] = None


class WizardOutlineRequest(BaseModel):
    """向导生成大纲请求（保留兼容性）"""
    title: str
    genre: Optional[str] = None
    description: str
    target_word_count: int = 100000
    chapter_count: int = 10
    reference_ids: Optional[List[int]] = None
    revision_request: Optional[str] = None
    current_outline: Optional[List[ChapterOutlineItem]] = None


class WizardCharactersRequest(BaseModel):
    """向导生成角色请求（保留兼容性）"""
    title: str
    genre: Optional[str] = None
    description: str
    outline: List[ChapterOutlineItem]


class WizardCreateRequest(BaseModel):
    """向导创建项目请求（保留兼容性）"""
    title: str
    genre: Optional[str] = None
    description: Optional[str] = None
    target_word_count: int = 100000
    outline: List[ChapterOutlineItem]
    characters: List[CharacterOutlineItem]
    reference_ids: Optional[List[int]] = None
    outline_text: Optional[str] = None


class WizardCreateResponse(BaseModel):
    """向导创建项目响应"""
    project_id: int
    message: str